# Exploring eBay Car Sales Data


In this project, the aim is to clean the data and analyze the used car listings.

I will work on the [Dataset: eBay](https://www.kaggle.com/orgesleka/used-cars-database/data) gathered from [eBay, Germany](https://www.ebay.de).
However, the dataset has been downsized to 50,000 rows from 371,000 but transformed to less-cleaned version of the data in kaggle.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import datetime as dt

In [2]:
# Read the data and assign to varible called autos using Latin-1 encoding
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In the dataset of autos, There are 50,000 rows and 20 columns containing 15 string and 5 integer types. Also, for some columns there are missing values (non have more than ~20% null values). There are some columns which contain dates and numbers stored as string.

I'll start by cleaning the column names to make the data easier to work with.

# Cleaning of Columns

First, we will convert the columns from __camelcase__ to __snake case__ and rename some columns to make them more descriptive.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
# Change the column names
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

I've change the column names as below to make them more meaningful and to make the analysis better and easier later.
* yearOfRegistration to registration_year
* monthOfRegistration to registration_month
* notRepairedDamage to unrepaired_damage
* dateCreated to ad_created
* nrOfPictures to num_photos

Other column names are converted to only snake case.

# Initial Data Exploration and Cleaning¶


In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

I initially observe that __seller__ and __offer_type__ column has almost one unique value. 
Also, __num_photos__ column looks odd having all 0 value so, I will investigate them further.

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

Obviously, I see that all the entries have 0 photo.

That's why I'll delete these three columns to have clearer data.

In [7]:
# Drop these three columns
autos = autos.drop(['num_photos', 'seller','offer_type'], axis=1 )

The other observation is __price__ and __odometer__ column whose entries are string instead of numerical. So, let's correct it.

In [8]:
# Price column from string to numeric values
autos['price'] = (autos['price']
                            .str.replace('$', '')
                            .str.replace(',', '')
                            .astype('int')
                            )
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [9]:
# Odometer column from string to numeric values
autos['odometer'] = (autos['odometer']
                            .str.replace('km', '')
                            .str.replace(',', '')
                            .astype(int)
                            )
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True) # adding km to name to indicate the unit
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

# Translation of German Words into English

The content of categorical data is in German. That's why, I will translate the german words in columns containing categorical data except __name__ column into English in order to understand the data better.

First, I will determine the German words that will be translated.

In [10]:
autos['vehicle_type'].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [11]:
autos['gearbox'].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [12]:
autos['fuel_type'].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [13]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [14]:
# Create dictionary for necessary columns to change values with mapping method
dictionary_vehicle_type = {'limousine':'limousine', 'kleinwagen':'hatchback', 'kombi':'station wagon','bus':'bus', 'cabrio':'cabrio','coupe':'coupe', 'suv':'suv', 'andere':'other'}
dictionary_gearbox = {'manuell':'manuel', 'automatik':'automatic'}
dictionary_fuel_type = {'benzin:':'gas', 'diesel':'diesel', 'lpg':'lpg', 'cng':'cng', 'hybrid':'hybrid', 'elektro':'electric', 'andere':'other'}
dictionary_unrepaired_damage = {'nein':'no', 'ja':'yes'}

# Perform translation by mapping method
autos['vehicle_type'] = autos['vehicle_type'].map(dictionary_vehicle_type)
autos['gearbox'] = autos['gearbox'].map(dictionary_gearbox)
autos['fuel_type'] = autos['fuel_type'].map(dictionary_fuel_type)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(dictionary_unrepaired_damage)

# Example of one column translated
print(autos['vehicle_type'].value_counts())


limousine        12859
hatchback        10822
station wagon     9127
bus               4093
cabrio            3061
coupe             2537
suv               1986
other              420
Name: vehicle_type, dtype: int64


# Extracting Car Model from Name

Car names were written by car owners that's why, the column values are not consistent to each other. However, in the name column, every word is separeted by \"_\" and the order of words are car brand and car model, respectively. Therefore, I will get second word as car model and create new column called \"model\".

In [15]:
# Extract car model from name column
autos['model'] = autos['name'].str.split('_', expand=True)[1]

# Exploring Price and Odometer 

In [16]:
# Statistics and information about price columns
print('The number of unique values of price is {} \n'.format(autos['price'].unique().shape[0]))
print('{} \n'.format(autos["price"].describe()))
print('The frequency of price values: \n Price    Frequency')
print(autos['price'].value_counts().sort_index(ascending=False))

The number of unique values of price is 2357 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

The frequency of price values: 
 Price    Frequency
99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             

According to info I get above, there are 1421 cars priced \\$0 which is about 2.8 \% of data which can be excluded. In addition to this, the maximum priced car is one hundred million dollars which seems a lot.

When I look at the frequency table, the are many cars priced under \\$50 as well as a few over \\$350,000.

Because eBay is an action site, it  might really have a bid starting by \\$1. That's why I will keep those entries however, I will remove the entries whose prices are over \\$350,000 as the prices in general increase steadily to \\$350,000 but then jump up to unrealistic prices.

In [17]:
# Drop rows containing prices $0 and over $350,000
autos = autos[autos['price'].between(1,350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [18]:
# Statistics and information about odometer columns
print('The number of unique values of odometer(km) is {} \n'.format(autos['odometer_km'].unique().shape[0]))
print('{} \n'.format(autos["odometer_km"].describe()))
print('The frequency of odometer values: \nodometer(km) Frequency')
print(autos['odometer_km'].value_counts().sort_index(ascending=False))

The number of unique values of odometer(km) is 13 

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

The frequency of odometer values: 
odometer(km) Frequency
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64


I clearly see that, the kilometers are rounded between min of 5000 km and max of 150,000 km. Also, the number of cars having more km are much more than those of having less km.

# Exploring Date Columns

Now, I will explore the columns containing dates; which are 
* __date_crawled__
* __last_seen__
* __ad_created__ 
* __registration_month__
* __registration_year__

I can differentiate these 5 columns based on whether eBay or crawler adds.
First two of them are added by the crawler and last three of them are added from eBay.

In [19]:
autos[['date_crawled', 'ad_created', 'last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
# The first 10 characters of date gives the date without time.
# used relative frequency then sorted by index
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

It looks like the site was crawled daily over a month of period __between March and April 2016__.
The distribution of date crawled is roughly uniform.

In [21]:
# The first 10 characters of date gives the date without time.
# used relative frequency then sorted by index
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Last seen date denotes that the last time a list was seen which allows me to indicate the car was sold and the list was removed. 

However, in the last three days, there are massive seen by crawlers (about 6-10 times of other values) in the listings which is more likely to do with crawling period ending but not with car sales.

In [22]:
# The first 10 characters of date gives the date without time.
# used relative frequency then sorted by index
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There are 76 unique dates in ad created date between June 2015 and April 2016. The oldest date is 9 months ago when a crawler visits the site. However, most of them are in 2 months period.

In [23]:
# Explore registration year of cars
print(autos['registration_year'].describe())
autos['registration_year'].value_counts().sort_index()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

The registration year of car denotes the age of the car.

According to the statistics about registration year of cars, there are odd values like min. value of 1000 which is way before cars were invented and max value of 9999 which is the date in the future. I will clean the inappropriate data.

# Dealing with Incorrect Registration Year Data

As a car can't be registered after the car list seen, these lists are definitely wrong data. Also, as I mentioned above, the odd values are also wrong data.

That's why, it's logical to choose car lists whose registration year is between 1900 and 2016. But before I drop the wrong data, I need to check the quantity of them if it won't damage the dataset.

In [24]:
# Check the percentage of the cars with registration year between 1900 and 2016
autos['registration_year'].between(1900, 2016).sum() / autos.shape[0]

0.961206630289303

as I can see above, the lists that I will drop cover only around 4\% of the dataset which is totally fine.

In [25]:
# Drop the list whose registarion year out of between 1900 and 2016
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
          ...   
1966    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1959    0.000129
1961    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.0000

It appears that most of the vehicles were registered in the past 20 years.

# Rough Calculation of the Time Passed until Cars Sold

I assume that last seen is the time when car sold. I also know the date when ad created. Therefore, the difference between them gives the time passed a car sold. Lastly, I will add this data to autos dataframe as a new column called time_sold.

However, I need to be careful when using this data, because it's not certin that last seen date is the date when car sold.

In [26]:
# Conversion from string to datetime
date_format = ("%Y-%m-%d %H:%M:%S") 
autos['last_seen'] = pd.to_datetime(autos['last_seen'], format=date_format)
autos['ad_created'] = pd.to_datetime(autos['ad_created'], format=date_format)

# Calculation of time passed until a car sold by using the assumption made above.
autos['time_sold'] = autos['last_seen'] - autos['ad_created']

# Convert to the time into dt.NaT if time is under a day
autos[autos['time_sold'] < dt.timedelta(days=1)] = pd.NaT
autos['time_sold'].value_counts(dropna=False).sort_index()

1 days 00:06:19         2
1 days 00:07:00         1
1 days 00:16:05         1
1 days 00:16:06         1
1 days 00:16:27         1
1 days 00:17:38         2
1 days 00:17:56         1
1 days 00:17:58         1
1 days 00:25:16         2
1 days 00:39:12         1
1 days 00:40:12         1
1 days 00:40:14         1
1 days 00:40:20         1
1 days 00:40:21         1
1 days 00:40:28         2
1 days 00:40:30         1
1 days 00:40:33         1
1 days 00:40:46         1
1 days 00:40:47         1
1 days 00:40:57         1
1 days 00:40:58         1
1 days 00:41:08         2
1 days 00:41:09         1
1 days 00:41:11         1
1 days 00:41:12         1
1 days 00:41:13         2
1 days 00:41:14         1
1 days 00:41:18         1
1 days 00:41:31         1
1 days 00:41:48         1
                     ... 
44 days 19:46:36        1
45 days 19:17:28        1
47 days 03:16:25        1
47 days 20:46:23        1
48 days 12:15:37        1
49 days 12:30:49        1
49 days 12:51:22        1
49 days 13:1

If I assume it as time passed to be sold, the maximum time a car waited to be sold is 300 days which is about six months. This is an average time to wait for a car to be sold.

However, it is more likely a car to be unsold yet if the car has less than 30 days because, there are plenty of cars in that range and under 30 days are not like to be ideal time for a car to be sold. Therefore, we don't know if they are sold or not.

Also, almost 14% of the cars have even no day between the time created and last seen. Therefore, there is no need to add this data to autos dataframe.

# Exploring Price by Brand

First, I will look at the relative frequeny table of brands and find the way out how we choose the brands that I will compare among.

In [27]:
# Explore common brands in the list
autos['brand'].value_counts(normalize=True)

volkswagen        0.208466
bmw               0.110873
opel              0.106223
mercedes_benz     0.097418
audi              0.085481
ford              0.069165
renault           0.047180
peugeot           0.029521
fiat              0.025443
seat              0.018230
skoda             0.016663
mazda             0.015992
nissan            0.015494
smart             0.014450
citroen           0.013952
toyota            0.012833
sonstige_autos    0.010371
hyundai           0.009973
volvo             0.009351
mini              0.009202
mitsubishi        0.008605
honda             0.008058
kia               0.007188
alfa_romeo        0.006839
porsche           0.006168
chevrolet         0.006044
suzuki            0.005919
chrysler          0.003731
daihatsu          0.002562
dacia             0.002537
subaru            0.002363
jeep              0.002363
land_rover        0.002213
saab              0.001716
jaguar            0.001567
daewoo            0.001442
trabant           0.001418
r

Looking at the table, I observe that German cars are the top on the list. Now, I will aggregate brands based on brands having frequency over 5\% in the list and find the mean price of each brand to compare.

In [28]:
# Extract common brands existing over 5% within the list.
brand_counts = autos['brand'].value_counts(normalize=True) 
common_brands = brand_counts[brand_counts > 0.05].index

# Create a dictionary containing common brands with their mean price in the list
brand_mean_price = {}
for brand in common_brands:
    selected_brand_data = autos[autos['brand'] == brand]
    mean_price = selected_brand_data['price'].mean()
    brand_mean_price[brand] = int(mean_price)
brand_mean_price

{'volkswagen': 5712,
 'bmw': 8624,
 'opel': 3132,
 'mercedes_benz': 9052,
 'audi': 9945,
 'ford': 3986}

The mean price of top 6 brands are not close to each other. 

The most and the least expensive brand are Audi and Opel, respectively.

Opel and Ford are less expensive where Mercedes Benz, Audi and BMW are more expensive among top brands.

Volkswagen is the mid-priced which might explain the popularity of the brand.


# Exploring Mileage by Brand


Now, for these top 6 brands, I will investigate average mileage and try to find any relation with their price distribution. 

Later on, I will create a series for each feature after that create 2 column dataframe in order to compare two features at the same time easily.

In [29]:
# Create a dictionary containing common brands with their mean mileage in the list
brand_mean_mileage = {}
for brand in common_brands:
    selected_brand_data = autos[autos['brand'] == brand]
    mean_mileage = selected_brand_data['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
brand_mean_mileage


{'volkswagen': 127280,
 'bmw': 131755,
 'opel': 128543,
 'mercedes_benz': 129951,
 'audi': 127768,
 'ford': 123180}

In [30]:
# creating series for brand mean price
bmp_series = pd.Series(brand_mean_price).sort_values(ascending=False)
bmp_series

audi             9945
mercedes_benz    9052
bmw              8624
volkswagen       5712
ford             3986
opel             3132
dtype: int64

In [31]:
# creating series for brand mean mileage
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
bmm_series

bmw              131755
mercedes_benz    129951
opel             128543
audi             127768
volkswagen       127280
ford             123180
dtype: int64

In [32]:
# creating dataframe for brand mean price and mileage
brand_info = pd.DataFrame(bmm_series, columns = ['mean_mileage'])
brand_info['mean_price'] = bmp_series
brand_info

mean_mileage  mean_price
bmw                  131755        8624
mercedes_benz        129951        9052
opel                 128543        3132
audi                 127768        9945
volkswagen           127280        5712
ford                 123180        3986

As I observe, there is no significant difference among mean mileage of top brands alike mean price. The highest difference in mean mileage is around 8000 km which might be assumed that the effect on price is almost none especially in this range of mileage. 

Also, there is a slight trend that more expensive cars tend do to have more mileage and vice versa.

# The Investigation of a Pattern between Mileage and Price

I will investigate the pattern between mileage and price within the brackets of 30,000 km until 150,000 km which is the highest mileage a car has in the list.

In [33]:
mean_price_vs_mileage_rage = {}
mean_price_vs_mileage_rage['0_30k_mile'] = int(autos.loc[autos['odometer_km'].between(0,30000), 'price'].mean())
mean_price_vs_mileage_rage['30_60k_mile'] = int(autos.loc[autos['odometer_km'].between(30001,60000), 'price'].mean())
mean_price_vs_mileage_rage['60_90k_mile'] = int(autos.loc[autos['odometer_km'].between(60001,90000), 'price'].mean())
mean_price_vs_mileage_rage['90_120k_mile'] = int(autos.loc[autos['odometer_km'].between(90001,120000), 'price'].mean())
mean_price_vs_mileage_rage['120_150k_mile'] = int(autos.loc[autos['odometer_km'].between(120001,150000), 'price'].mean())
mean_price_vs_mileage_rage


{'0_30k_mile': 15249,
 '30_60k_mile': 13989,
 '60_90k_mile': 9761,
 '90_120k_mile': 8414,
 '120_150k_mile': 4311}

Clearly, there is a decreasing trend in average price while mileage increses. The average price of [0-30] bracket is almost 4 times less than the average price of [120-150] bracket 

# The Exploration of Average Price depend on Unrepaired Damage

I will investigate how average price differs whether a car has unrepaired damage or not. First, I will check the column variables, then calculate the average price according to its category.

To get more meaningful insights, I will just analyze last 20 years with 5 years brackets because the difference between the price of 20 year old car and the price of same car with unrepaired damage is not expected to be significant.

In [34]:
autos['unrepaired_damage'].value_counts(dropna=False)


no     29732
NaN    13324
yes     3625
Name: unrepaired_damage, dtype: int64

There are 13324 undefined values in the unrepaired damage category which is actually 28.5\% of values. However, it will be still a good insight although I will exclude NaN values in the analysis.

In [35]:
# Splitting cars having unrepaired damage and no damage based on 4 "5 year"-brackets car age
mp_unrepaired_damage = {}
mp_no_unrepaired_damage = {}
n = 2016
for t in range(4):
    name = str(n - 4) + ' - ' + str(n) # row names in dataframe to be created
    mp_unrepaired_damage[name] = autos.loc[(autos['unrepaired_damage'] == 'yes') & (autos['registration_year'].between(n-4, n)), 'price'].mean()
    mp_no_unrepaired_damage[name] = autos.loc[(autos['unrepaired_damage'] == 'no') & (autos['registration_year'].between(n-4, n)), 'price'].mean()
    n = n - 5

mp_unrepaired_series = pd.Series(mp_unrepaired_damage) # Create Series from car with damage dictionary
mp_no_unrepaired_series = pd.Series(mp_no_unrepaired_damage) # Create Series from car with no damage dictionary
price_vs_damage = pd.DataFrame(mp_unrepaired_series) # Create DataFrame from Series created
price_vs_damage.columns = ['have_unrepaired_damage']
price_vs_damage['have_no_unrepaired_damage'] = mp_no_unrepaired_series # adding second column to dataframe

price_vs_damage

have_unrepaired_damage  have_no_unrepaired_damage
2012 - 2016             6270.787565               18429.598912
2007 - 2011             6835.195822               11289.108835
2002 - 2006             2605.852174                5232.816563
1997 - 2001              995.590562                2403.994219

According to the table above, the mean price of the cars having unrepaired damage is almost a half of the mean price of the cars having no unrepaired damage in each year except the years between 2012- 2016. Between 2012 - 2016 there is a price difference almost three times between the two categories. I might explain this by as the car getting older, the price change due to unrepaired damage has less effect because its price has already begun to drop because of its mileage and age.

Also, the mean price of car having unrepaired damage between 2007 - 2011 is more than the mean price of car having unrepaired damage between 2012 - 2016 which is odd. However, in the dataset, it is known that if there is a unrepaired damage or not but not the type or the cost of the damage. That's I cannot do any further analysis for the damage cost affecting the car price. Therefore, this might explain the distribution of mean price of cars having unrepaired damage because, regardless of mileage, age or any feature, the car might have a serious damage needed to be repaired which affects its price significantly.

Moreover, I observe that the price drops in each category when the car age increases. it is seen that for the cars having no unrepaired damage has the mean price drop of around 1/2 of its price in every 5 years.  